contracts/cw721-expiration/src/query.rs

```diff
@@ -1,296 +1,347 @@
-        include_invalid: bool,
+        include_expired_nft: bool,
     ) -> StdResult<TokensResponse> {
-        let tokens = self.base_contract.tokens(deps, owner, start_after, limit)?;
-        if include_invalid {
+        let tokens =
+            self.base_contract
+                .query_tokens(deps, env.clone(), owner, start_after, limit)?;
+        if include_expired_nft {
             return Ok(tokens);
         }
         let filtered: Vec<_> = tokens
             .tokens
             .iter()
             .filter(|token_id| {
-                self.is_valid_nft(deps, &env, token_id).unwrap_or(false) // Convert Option<bool> to bool
+                self.is_nft_expired(deps, &env, token_id).unwrap_or(false) // Convert Option<bool> to bool
             })
             .map(|token_id| token_id.to_string())
             .collect();
         Ok(TokensResponse { tokens: filtered })
     }
 
-    pub fn all_tokens(
+    pub fn query_all_tokens_include_expired_nft(
         &self,
         deps: Deps,
         env: Env,
         start_after: Option<String>,
         limit: Option<u32>,
-        include_invalid: bool,
+        include_expired_nft: bool,
     ) -> Result<TokensResponse, ContractError> {
-        let tokens = self.base_contract.all_tokens(deps, start_after, limit)?;
-        if include_invalid {
+        let tokens = self
+            .base_contract
+            .query_all_tokens(deps, env.clone(), start_after, limit)?;
+        if include_expired_nft {
             return Ok(tokens);
         }
         let filtered: Vec<_> = tokens
             .tokens
             .iter()
             .filter(|token_id| {
-                self.is_valid_nft(deps, &env, token_id).unwrap_or(false) // Convert Option<bool> to bool
+                self.is_nft_expired(deps, &env, token_id).unwrap_or(false) // Convert Option<bool> to bool
             })
             .map(|token_id| token_id.to_string())
             .collect();
         Ok(TokensResponse { tokens: filtered })
     }
 
-    pub fn all_nft_info(
+    pub fn query_all_nft_info_include_expired_nft(
         &self,
         deps: Deps,
         env: Env,
         token_id: String,
-        include_expired: bool,
-        include_invalid: bool,
-    ) -> Result<AllNftInfoResponse<Extension>, ContractError> {
-        if !include_invalid {
-            self.assert_valid_nft(deps, &env, token_id.as_str())?;
+        include_expired_approval: bool,
+        include_expired_nft: bool,
+    ) -> Result<AllNftInfoResponse<TMetadataExtension>, ContractError> {
+        if !include_expired_nft {
+            self.assert_nft_expired(deps, &env, token_id.as_str())?;
         }
         Ok(self
             .base_contract
-            .all_nft_info(deps, env, token_id, include_expired)?)
+            .query_all_nft_info(deps, env, token_id, include_expired_approval)?)
+    }
+
+    // --- helpers ---
+    pub fn is_nft_expired(&self, deps: Deps, env: &Env, token_id: &str) -> StdResult<bool> {
```

Comment:
It seems like this "expired nft" is the bulk of the PR. What exactly is an "expired nft?"

Reply:
Example: you provided subscriptions, like a user can subscribe a domain/node/any service for 12 months. Now you user can buy this service, by buying an NFT which expires after 12 months. So creator can create a platform, where user can approve/have access by this NFT, once it got expired, subscription ends.